ts/parsers/paragraphparser.ts

```diff
@@ -1,37 +1,155 @@
+                text: txtElement["a:t"] || "",
+                textCharacterProperties: this.determineTextProperties(
+                    getValueAtPath(txtElement, '["a:rPr"][0]')
+                )
+            };
+        });
+
+        return {
+            content: content,
+            paragraphProperties: this.determineParagraphProperties(paragraph)
+        };
+    }
+
+    public static getListlevel(paragraph): number {
+        const level = getValueAtPath(paragraph, '["a:pPr"][0]["$"]["lvl"]');
+
+        return level ? parseInt(level) : 0;
+    }
+
+    public static getListType(paragraph): ListType {
+        if (checkPath(paragraph, '["a:pPr"][0]["a:buAutoNum"]')) {
+            return ListType.Ordered;
+        }
+
+        return ListType.UnOrdered;
+    }
+
+    public static restructureList(list: List): List {
+        for (let i = 0; i < list.listItems.length - 1; i++) {
+            if (list.listItems[i + 1].list) {
+                list.listItems[i]["list"] = this.restructureList(list.listItems[i + 1].list);
+                list.listItems.splice(i + 1, 1);
+            }
+        }
+        return list;
+    }
+
     public static extractParagraphElements(paragraphs: any[]): PowerpointElement["paragraph"] {
         if (!paragraphs || paragraphs.length === 0) {
             return null;
         }
 
-        return paragraphs.map((paragraph) => {
-            const textElements = paragraph["a:r"] || [];
-            const content = textElements.map((txtElement) => {
-                return {
-                    text: txtElement["a:t"] || "",
-                    textCharacterProperties: this.determineTextProperties(getValueAtPath(txtElement, '["a:rPr"][0]'))
-                };
-            });
+        const paras = [];
+        const stack = [];
+        const paragraph: Paragraph = {
+            list: {
+                listType: ListType.Ordered,
+                listItems: []
+            }
+        };
+        let currentParagraph = paragraph;
+        let currentLevel = -1;
 
-            return {
-                content: content,
-                paragraphProperties: this.determineParagraphProperties(paragraph)
-            };
-        });
+        for (const p of paragraphs) {
+            if (this.isList(p)) {
+                const listLevel = this.getListlevel(p);
+
+                //if its the first of the list kind
+                if (currentLevel === -1) {
+                    currentLevel = 0;
+                    currentParagraph.list.listType = this.getListType(p);
+                    currentParagraph.list.listItems.push(this.getParagraph(p));
+                    stack.push(currentParagraph);
+                }
+                //if the level is same keep pushing the list items in the same array
+                else if (listLevel === currentLevel) {
+                    currentParagraph.list.listItems.push(this.getParagraph(p));
+                } else if (listLevel > currentLevel) {
```

Comment:
don't need else's .. the current level will be either bigger, equal to or less than list level so it will always only match one of three `if` statements even if you don't put else.

Reply:
It was like that before, but if statement before these ones, increments the current level, therefore it results in going it to the next if statement.

Reply:
ah yeah will need to break early, good point .. fine. Ignore.